File: src/shared/services/metrics/metrics.go
/*
 * Copyright 2018- The Pixie Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package metrics

import (
	"net/http"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/collectors"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

// mux is an interface describing the methods InstallHandler requires.
type mux interface {
	Handle(pattern string, handler http.Handler)
}

// MustRegisterMetricsHandler registers the metrics endpoint.
func MustRegisterMetricsHandler(mux mux) {
	mux.Handle("/metrics", promhttp.HandlerFor(
		prometheus.DefaultGatherer,
		promhttp.HandlerOpts{
			// Opt into OpenMetrics to support exemplars.
			EnableOpenMetrics: true,
		},
	))
	prometheus.MustRegister(collectors.NewBuildInfoCollector())
}

// MustRegisterMetricsHandlerNoDefaultMetrics registers the metrics endpoint but removes all of the default metrics that prometheus includes, so only custom metrics are collected and reported.
func MustRegisterMetricsHandlerNoDefaultMetrics(mux mux) {
	mux.Handle("/metrics", promhttp.HandlerFor(
		prometheus.DefaultGatherer,
		promhttp.HandlerOpts{
			// Opt into OpenMetrics to support exemplars.
			EnableOpenMetrics: true,
		},
	))

	// Unregister the default metrics the prometheus golang client provides.
	prometheus.Unregister(collectors.NewGoCollector())
	prometheus.Unregister(collectors.NewProcessCollector(collectors.ProcessCollectorOpts{}))
}

File: src/stirling/protocol_inference/utils/metrics.py
# Copyright 2018- The Pixie Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

import torch


class MeanPerClass:
    def __init__(self, num_classes):
        self.confusion_matrix = torch.zeros(num_classes, num_classes)

    def update(self, targets, preds):
        for t, p in zip(targets.flatten(), preds.flatten()):
            self.confusion_matrix[t.item(), p.item()] += 1

    def avg(self):
        return torch.mean(
            self.confusion_matrix.diag() / (self.confusion_matrix.sum(1) + 1e-8)).item()

    def __str__(self):
        return "MeanPerClass"
